pkg/api/validation/validation_test.go

```diff
@@ -0,0 +1,163 @@
+			}
+		}
+
+		// check for illegal contents
+		for _, illegalContent := range api.NameMayNotContain {
+			illegalName := "a" + illegalContent + "b"
+
+			apiObjectMeta.Set(reflect.ValueOf(kapi.ObjectMeta{Name: illegalName}))
+
+			errList := validationInfo.validator.Validate(apiValue.Interface().(runtime.Object))
+			_, requiredMessage := api.MinimalNameRequirements(illegalName, false)
+
+			if len(errList) == 0 {
+				t.Errorf("expected error for %v in %v not found amongst %v", illegalName, apiType.Elem(), errList)
+				continue
+			}
+
+			foundExpectedError := false
+			for _, err := range errList {
+				validationError, ok := err.(*fielderrors.ValidationError)
+				if !ok || validationError.Type != fielderrors.ValidationErrorTypeInvalid || validationError.Field != "metadata.name" {
+					continue
+				}
+
+				if validationError.Detail == requiredMessage {
+					foundExpectedError = true
+					break
+				}
+				if validationError.Detail == validation.DNSSubdomainErrorMsg {
+					foundExpectedError = true
+					break
+				}
+			}
+
+			if !foundExpectedError {
+				t.Errorf("expected error for %v in %v not found amongst %v", illegalName, apiType.Elem(), errList)
+			}
+		}
+	}
+}
+
+func TestObjectMeta(t *testing.T) {
+	for apiType, validationInfo := range Validator.typeToValidator {
+		if !validationInfo.hasObjectMeta {
+			continue
+		}
+
+		apiValue := reflect.New(apiType.Elem())
+		apiObjectMeta := apiValue.Elem().FieldByName("ObjectMeta")
+
+		if validationInfo.isNamespaced {
+			apiObjectMeta.Set(reflect.ValueOf(kapi.ObjectMeta{Name: getValidName(apiType)}))
+		} else {
+			apiObjectMeta.Set(reflect.ValueOf(kapi.ObjectMeta{Name: getValidName(apiType), Namespace: kapi.NamespaceDefault}))
+		}
+
+		// these have special case conversions that always strip
+		switch apiValue.Interface().(type) {
+		case *authorizationapi.ClusterPolicyBinding, *authorizationapi.PolicyBinding, *authorizationapi.ClusterPolicy, *authorizationapi.Policy:
+			continue
+		}
+
+		errList := validationInfo.validator.Validate(apiValue.Interface().(runtime.Object))
+		requiredErrors := validation.ValidateObjectMeta(apiObjectMeta.Addr().Interface().(*kapi.ObjectMeta), validationInfo.isNamespaced, api.MinimalNameRequirements).Prefix("metadata")
+
+		if len(errList) == 0 {
+			t.Errorf("expected errors %v in %v not found amongst %v", requiredErrors, apiType.Elem(), errList)
+			continue
+		}
+
+		for _, requiredError := range requiredErrors {
+			foundExpectedError := false
+
+			for _, err := range errList {
+				validationError, ok := err.(*fielderrors.ValidationError)
+				if !ok {
+					continue
+				}
+
+				if fmt.Sprintf("%v", validationError) == fmt.Sprintf("%v", requiredError) {
+					foundExpectedError = true
+					break
+				}
+			}
+
+			if !foundExpectedError {
+				t.Errorf("expected error %v in %v not found amongst %v", requiredError, apiType.Elem(), errList)
```

Comment:
add a message explaining the likely fix ("You need to call `ValidateObjectMeta` in your validator") or people will start copy/pasting the error message we want into their custom validators